### PR TITLE
Ensure that iast json tag does not exceed the maximum allowed

### PR DESCRIPTION
**Pull request by jandro996**
# What Does This Do

Verify that serialized vulnerabilities not exceed 25Kb, and in case it is exceeded, it serialize the vulnerabilities again by removing the sources and replacing the evidence with a generic value of “MAX SIZE EXCEEDED”

# Motivation

The span tags are limited to 25 kb, so we need to ensure that the iast json tag does not exceed this limit when serializing the vulnerabilities.

# Additional Notes
Jira ticket: [APPSEC-11517]

[APPSEC-11517]: https://datadoghq.atlassian.net/browse/APPSEC-11517?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/TruncatedVulnerabilitiesAdapter.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      writer.beginArray();
+      for (Vulnerability vulnerability : vulnerabilities) {
+        vulnerabilityAdapter.toJson(writer, vulnerability);
+      }
+      writer.endArray();
+    }
+    writer.endObject();
+  }
+
+  public static class TruncatedVulnerabilityAdapter extends FormattingAdapter<Vulnerability> {
```
Comment (cataphract) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/TruncatedVulnerabilitiesAdapter.java:46]: could be private

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/TruncatedVulnerabilitiesAdapter.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      evidenceAdapter.toJson(writer, value.getEvidence());
+      writer.name("hash");
+      writer.value(value.getHash());
+      writer.name("location");
+      locationAdapter.toJson(writer, value.getLocation());
+      writer.endObject();
+    }
+  }
+
+  public static class TruncatedEvidenceAdapter extends FormattingAdapter<Evidence> {
```
Comment (cataphract) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/TruncatedVulnerabilitiesAdapter.java:79]: idem

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/VulnerabilityEncoding.java [hunk 3]
```diff
@@ -13,6 +14,14 @@ public class VulnerabilityEncoding {
...
-    return BATCH_ADAPTER.toJson(value);
+    String json = BATCH_ADAPTER.toJson(value);
+    return json.getBytes().length > 25000
+        ? getExceededTagSizeJson(new TruncatedVulnerabilities(value.getVulnerabilities()))
+        : json;
+  }
+
+  static String getExceededTagSizeJson(final TruncatedVulnerabilities truncatedVulnerabilities) {
+    // TODO report via telemetry
+    return MOSHI.adapter(TruncatedVulnerabilities.class).toJson(truncatedVulnerabilities);
```
Comment (cataphract) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/VulnerabilityEncoding.java:25]: `MOSHI.adapter(TruncatedVulnerabilities.class)` is relatively expensive and should be put in a static final.

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/TruncatedVulnerabilitiesAdapter.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public void toJson(@Nonnull JsonWriter writer, @Nullable Evidence evidence) throws IOException {
+      writer.beginObject();
+      writer.name("value");
+      writer.value(MAX_SIZE_EXCEEDED);
+      writer.endObject();
+    }
+  }
+
+  public static class TruncatedVulnerabilities {
```
Comment (cataphract) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/TruncatedVulnerabilitiesAdapter.java:89]: I'd reverse the enclosing/enclosed classes (make `TruncatedVulnerabilities` the top level class instead). Or just move `TruncatedVulnerabilities` to the top level.

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/VulnerabilityEncoding.java [hunk 5]
```diff
@@ -12,7 +12,18 @@ public class VulnerabilityEncoding {
   private static final JsonAdapter<VulnerabilityBatch> BATCH_ADAPTER =
       MOSHI.adapter(VulnerabilityBatch.class);
 
+  private static final JsonAdapter<TruncatedVulnerabilities> TRUNCATED_VULNERABILITIES_ADAPTER =
```
Comment (manuel-alvarez-alvarez) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/VulnerabilityEncoding.java:15]: The new truncated adapter can be for `VulnerabilityBatch` so you don't need an extra DTO, or am I missing something?

Comment (jandro996) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/VulnerabilityEncoding.java:15]: Thera are a few factors that have made me opt for that solution:

- We can't define different adapters for the same class using the current adapter factory. 
- We need to evaluate the tag size in the VulnerabilityEncoding class as we don't have the final json earlier
- Use the context with a new property to discern between regular and truncated adapters could be an option but it's not accesible outside the adapter factory

Taking this into account, it seems "cleaner" the wrapper approach than maintain another factory with truncation adapters

**Review by smola — APPROVED** [review]

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/VulnerabilityEncoding.java [hunk 6]
```diff
@@ -12,7 +12,18 @@ public class VulnerabilityEncoding {
...
+      MOSHI.adapter(TruncatedVulnerabilities.class);
+
   public static String toJson(final VulnerabilityBatch value) {
-    return BATCH_ADAPTER.toJson(value);
+    String json = BATCH_ADAPTER.toJson(value);
+    return json.getBytes().length > 25000
```
Comment (manuel-alvarez-alvarez) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/model/json/VulnerabilityEncoding.java:20]: You can extract a constant here to give a name to this magic number

**Review by manuel-alvarez-alvarez — APPROVED** [review]

**Pull request merged**